### PR TITLE
Timestepping plugin function

### PR DESCRIPTION
**Pull request by tjhei**
- add a "function" plugin and a test for it
    
Part 3 of N for #3549

### CODE REVIEW

**Comment by tjhei** [comment]
ready for review.

File: source/time_stepping/function.cc [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+}
+
+// explicit instantiations
+namespace aspect
+{
+  namespace TimeStepping
+  {
+    ASPECT_REGISTER_TIME_STEPPING_MODEL(Function,
+                                        "function",
+                                        "This model uses a time step specified in the parameter file.")
```
Comment (MFraters) [inline source/time_stepping/function.cc:102]: could you indicate that is is specified by a function and that it uses the advance reaction?

File: tests/time_stepping_function.prm [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+subsection Time stepping
+
+  set List of model names = function
+  subsection Function
+    set Function expression = if (time<5e3,2e2+1e2*cos(2*3.141*time/5e3),3e2+time/1e3)
+  end
+end
+set Maximum relative increase in time step     = 2147483647
+
+
```
Comment (MFraters) [inline tests/time_stepping_function.prm:21]: why so many newlines here?

File: source/time_stepping/interface.cc [hunk 3]
```diff
@@ -108,7 +108,13 @@ namespace aspect
         new_time_step = std::min(new_time_step, this->get_parameters().maximum_first_time_step);
 
       // Make sure we reduce the time step length appropriately if we terminate after this step
-      return termination_manager.check_for_last_time_step(new_time_step);
+      new_time_step = termination_manager.check_for_last_time_step(new_time_step);
+
+      AssertThrow (new_time_step > 0,
+                   ExcMessage("The time step length for the each time step needs to be positive, "
+                              "but the computed step length was: " + std::to_string(new_time_step) + ". "
+                              "Please check the time stepping plugins and for non-positive material properties."));
```
Comment (gassmoeller) [inline source/time_stepping/interface.cc:116]: Something is off in this line `and ...? for`. Looks good otherwise except for Menno's comments.

**Pull request merged**